format/ldb/ldb.go

```diff
@@ -0,0 +1,279 @@
+		}
+		dataHandles = append(dataHandles, handle)
+	}, d)
+
+	// meta
+
+	if len(metaHandles) > 0 {
+		d.FieldArray("meta", func(d *decode.D) {
+			for _, handle := range metaHandles {
+				d.SeekAbs(int64(handle.Offset) * 8)
+				fieldStructBlock("meta_block", int64(handle.Size), readMetaContent, nil, d)
+			}
+		})
+	}
+
+	// data
+
+	if len(dataHandles) > 0 {
+		d.FieldArray("data", func(d *decode.D) {
+			for _, handle := range dataHandles {
+				d.SeekAbs(int64(handle.Offset) * 8)
+				fieldStructBlock("data_block", int64(handle.Size), readKeyValueContent, nil, d)
+			}
+		})
+	}
+
+	return nil
+}
+
+// Helpers
+
+func fieldStructBlock(name string, size int64, readBlockContent func(size int64, valueCallbackFn func(d *decode.D), d *decode.D), valueCallbackFn func(d *decode.D), d *decode.D) *decode.D {
+	// ReadBlock: https://github.com/google/leveldb/blob/main/table/format.cc#L69
+	return d.FieldStruct(name, func(d *decode.D) {
+		start := d.Pos()
+		br := d.RawLen(size * 8)
+		compressionType := d.FieldU8("compression", compressionTypes, scalar.UintHex)
+		// validate crc
+		data := d.ReadAllBits(br)
+		bytesToCheck := append(data, uint8(compressionType))
+		maskedCRCInt := maskedCrc32(bytesToCheck)
+		d.FieldU32("crc", d.UintAssert(uint64(maskedCRCInt)), scalar.UintHex)
+
+		d.SeekAbs(start)
+		if compressionType == compressionTypeNone {
+			d.FieldStruct("uncompressed", func(d *decode.D) {
+				readBlockContent(size, valueCallbackFn, d)
+			})
+		} else {
+			compressedSize := size
+			compressed := data
+			bb := &bytes.Buffer{}
+			_ = bb
+			switch compressionType {
+			case compressionTypeSnappy:
+				decompressed, err := snappy.Decode(nil, compressed)
+				if err != nil {
+					d.Fatalf("failed decompressing data: %v", err)
+				}
+				d.Copy(bb, bytes.NewReader(decompressed))
+			default:
+				d.Fatalf("Unsupported compression type: %x", compressionType)
+			}
+			d.FieldStructRootBitBufFn("uncompressed", bitio.NewBitReader(bb.Bytes(), -1), func(d *decode.D) {
+				readBlockContent(int64(bb.Len()), valueCallbackFn, d)
+			})
+			d.FieldRawLen("compressed", compressedSize*8)
+		}
+
+	})
+}
+
+func readKeyValueContent(size int64, valueCallbackFn func(d *decode.D), d *decode.D) {
+	// https://github.com/google/leveldb/blob/main/table/block_builder.cc#L16
+	// https://github.com/google/leveldb/blob/main/table/block.cc
+	uint32Size := int64(32)
+	uint64Size := int64(64)
+	start := d.Pos()
+	end := start + size*8
+
+	var restartOffset int64
+	d.SeekAbs(end - uint32Size)
+	d.FieldStruct("trailer", func(d *decode.D) {
+		numRestarts := int64(d.FieldU32("num_restarts"))
+		restartOffset = size*8 - (1+numRestarts)*uint32Size
+		d.SeekAbs(start + restartOffset)
+		d.FieldArray("restarts", func(d *decode.D) {
+			for i := 0; i < int(numRestarts); i++ {
+				d.FieldU32("restart")
+			}
+		})
+	})
+	// TK: how do you make an empty entries-array appear _above_ the trailer?
+	// Right now, its omited if empty.
```

Comment:
Typically a key-value block has (1) entries and (2) trailer. However, occasionally there's only a trailer and no entries.
What to do in such scenarios?
Option 1: leave away the entries array all together. (Current solution.)
Option 2: Show an empty entries array. (Proposition. However, it ends up being shown below the trailer, since it's empty. We need to read the trailer first, to figure out if there's an entries array or not.)

Reply:
I guess it would be nice to have an empty entries array for clarity? i will have to look closer at how to solve this. But don't think it's a blocker to merge, can be changed/fixed later.